api/service/pprof/service.go

```diff
@@ -0,0 +1,235 @@
+}
+
+type Profile struct {
+	Name     string
+	Interval int
+	Debug    int
+}
+
+func (p Config) String() string {
+	return fmt.Sprintf("%v, %v, %v, %v/%v/%v", p.Enabled, p.ListenAddr, p.Folder, p.ProfileNames, p.ProfileIntervals, p.ProfileDebugValues)
+}
+
+// Constants for profile names
+const (
+	CPU = "cpu"
+)
+
+// Service provides access to pprof profiles via HTTP and can save them to local disk periodically as user settings.
+type Service struct {
+	config   Config
+	profiles map[string]Profile
+}
+
+var (
+	initOnce sync.Once
+	svc      = &Service{}
+)
+
+// NewService creates the new pprof service
+func NewService(cfg Config) *Service {
+	initOnce.Do(func() {
+		svc = newService(cfg)
+	})
+	return svc
+}
+
+func newService(cfg Config) *Service {
+	if !cfg.Enabled {
+		utils.Logger().Info().Msg("Pprof service disabled...")
+		return nil
+	}
+
+	utils.Logger().Debug().Str("cfg", cfg.String()).Msg("Pprof")
+	svc.config = cfg
+
+	if profiles, err := cfg.unpackProfilesIntoMap(); err != nil {
+		utils.Logger().Error().Err(err).Msg("Could not unpack profiles into map")
+		return nil
+	} else {
+		svc.profiles = profiles
+	}
+
+	go func() {
+		utils.Logger().Info().Str("address", cfg.ListenAddr).Msg("Starting pprof HTTP service")
+		http.ListenAndServe(cfg.ListenAddr, nil)
+	}()
+
+	return svc
+}
+
+// Start start the service
+func (s *Service) Start() error {
+	dir := getBasePath(s.config.Folder)
+	err := os.MkdirAll(dir, os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+
+	if cpuProfile, ok := s.profiles[CPU]; ok {
+		handleCpuProfile(cpuProfile, dir, false)
+	}
+
+	for _, profile := range s.profiles {
+		scheduleProfile(profile, dir)
+	}
+
+	return nil
+}
+
+// Stop stop the service
+func (s *Service) Stop() error {
+	dir := getBasePath(s.config.Folder)
+	for _, profile := range s.profiles {
+		if profile.Name == CPU {
+			pprof.StopCPUProfile()
+		} else {
+			saveProfile(profile, dir)
+		}
+	}
+	return nil
+}
+
+// APIs return all APIs of the service
+func (s *Service) APIs() []rpc.API {
+	return nil
+}
+
+// scheduleProfile schedules the provided profile based on the specified interval (e.g. saves the profile every x seconds)
+func scheduleProfile(profile Profile, dir string) {
+	go func() {
+		if profile.Interval > 0 {
+			ticker := time.NewTicker(time.Second * time.Duration(profile.Interval))
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					if profile.Name == CPU {
+						handleCpuProfile(profile, dir, true)
+					} else {
+						saveProfile(profile, dir)
+					}
+				}
+			}
+		}
+	}()
+}
+
+// saveProfile saves the provided profile in the specified directory
+func saveProfile(profile Profile, dir string) {
+	f, err := newTempFile(dir, profile.Name, ".pb.gz")
+	if err == nil {
+		defer f.Close()
+		if p := pprof.Lookup(profile.Name); p == nil {
+			utils.Logger().Error().Msg(fmt.Sprintf("Profile does not exist: %s", profile.Name))
+		} else {
+			runtime.GC() // get up-to-date statistics
```

Comment:
The default example suggests to call garbage collection as described here https://golang.org/pkg/runtime/pprof/. But I agree it could potentially be expensive -> removed. Maybe making the call optional with an additional flag?